backupccl, changefeedccl, importer: log (sanitized) destination URIs

alyshanjahani-crl opened the pull request:
Fixes: https://github.com/cockroachdb/cockroach/issues/86408
Epic: https://cockroachlabs.atlassian.net/browse/CC-5615

Previously, there were no logs that indicated the destinations CRDB was attempting to connect to during egress enabled jobs (BACKUP,IMPORT, CHANGEFEED, RESTORE).

This needed to change because security teams want the ability to analyze egress traffic coming out of a CRDB cluster. These logs allow them to determine which traffic is actually user initiated (network egress logs align with CRDB logs). And furthermore, if that traffic is initiated by a malicious user (comparing with historical logs indicating unexpected destinations).

To address this, jobs that leverage egress have been modified to include additional logging. Note that we mark the logged destinations as redact safe since we ensure to log only sanitized URIs.
While there may be other ways to manipulate CRDB to make egress calls, these jobs (BACKUP, IMPORT, CHANGEFEED, RESTORE) are the most likely targets for egress manipulation.

Release Note (general change): Bulk operations now log the (sanitized) destinations they are connecting to. For example "backup planning to connect to destination gs://test/backupadhoc?AUTH=specified&CREDENTIALS=redacted"

alyshanjahani-crl committed:
backupccl, changefeedccl, importer: log (sanitized) destination URIs

Previously, there were no logs that indicated the destinations CRDB was
attempting to connect to during egress enabled jobs (BACKUP,IMPORT,
CHANGEFEED, RESTORE).

This needed to change because security teams want the ability to analyze
egress traffic coming out of a CRDB cluster. These logs allow them to
determine which traffic is actually user initiated (network egress logs
align with CRDB logs). And furthermore, if that traffic is initiated by
a malicious user (comparing with historical logs indicating unexpected
destinations).

To address this, jobs that leverage egress have been modified to include
additional logging. Note that we mark the logged destinations as redact
safe since we ensure to log only sanitized URIs.
While there may be other ways to manipulate CRDB to make egress calls,
these jobs (BACKUP, IMPORT, CHANGEFEED, RESTORE) are the most likely
targets for egress manipulation.

Release Note (general change): Bulk operations now log the (sanitized)
destinations they are connecting to. For example "backup planning to
connect to destination gs://test/backupadhoc?AUTH=specified&CREDENTIALS=redacted"

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89531)
<!-- Reviewable:end -->

stevendanna reviewed (commented):
My understanding is that the goal here is to provide a low-cost way to line up other logs related to external access with specific user actions.

In that regard, I'll note a few things you might want to take a look at:

1) For the new external-connection features, the current code will only log the name of the external connection.

2) You probably want to modify EXPORT as well.  There are a few other functions such as `crdb_internal.write_file` and `crdb_internal.read_file` that you could do as well, but those aren't used very often.

Even though they are single log-lines, it might be nice to move this logging into a well named function so that it doesn't get accidentally removed -- especially for the log lines in restore, which currently live in functions that aren't necessarily obvious would be logging.

alyshanjahani-crl commented:
> My understanding is that the goal here is to provide a low-cost way to line up other logs related to external access with specific user actions.

Correct 👌 

> Even though they are single log-lines, it might be nice to move this logging into a well named function so that it doesn't get accidentally removed -- especially for the log lines in restore, which currently live in functions that aren't necessarily obvious would be logging.

Done.

> For the new external-connection features, the current code will only log the name of the external connection.
What are the "new external-connection" features? Are these things other than the BACKUP, RESTORE, IMPORT, EXPORT, CREATE CHANGEFEED statements ?


> You probably want to modify EXPORT as well
Could you point me to where that would be? I started this PR by finding the usages of `cloud.SanitizeExternalStorageURI` and that did not lead me to EXPORT. I also don't see an `export_planning` file similar to how i see for the other operations.

dt commented:
This is currently logging only in planning, right? So if someone starts a changefeed -- which would log -- and then pauses it, and then several days later comes back and resumes it, it'll start making network connections at that point, but will not log anything new at that point, since it doesn't go back through planning on resume. Is that ok?

alyshanjahani-crl commented:
>  it'll start making network connections at that point, but will not log anything new at that point, since it doesn't go back through planning on resume. Is that ok?

Yes that's fine, give that it is not possible to change the destination by pausing/resuming.

stevendanna commented:
Another note is that this won't cover confluent schema registries (CDC) and KMS (backup/restore).

catj-cockroach reviewed (approved):
This looks good to me! Are these logs going to the Ops channel?

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89531#-:-NFxxojTACnEoBP0hW-v:bhuktsi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl and @benbardin)



<!-- Sent from Reviewable.io -->

alyshanjahani-crl commented:
@stevendanna I've added logging around the creation of external connections which satisfies our use case (since external connections currently can't be altered/modified - if they did we would want logging there too).

I also added logging for EXPORT and backup/restore KMS.

For confluent schema registries i could find a way to easily do this only on changefeed set up only. However, like some other less common egress scenarios we are willing to ignore logging here.

PTAL

alyshanjahani-crl commented:
> This looks good to me! Are these logs going to the Ops channel?

They were going to Dev. I have changed the logging statements to send to the Ops channel.

alyshanjahani-crl commented on pkg/sql/importer/exportcsv_test.go:
```diff
@@ -451,11 +451,11 @@ func TestExportFeatureFlag(t *testing.T) {
 	sqlDB.Exec(t, `SET CLUSTER SETTING feature.export.enabled = FALSE`)
 	sqlDB.Exec(t, `CREATE TABLE feature_flags (a INT PRIMARY KEY)`)
 	sqlDB.ExpectErr(t, `feature EXPORT was disabled by the database administrator`,
-		`EXPORT INTO CSV 'nodelocal://0/%s/' FROM TABLE feature_flags`)
```
I'm not sure how this was passing previously.. 

After my introduction of logging the url in`createPlanForExport` this test started failing w/ 
```
    exportcsv_test.go:458: error executing 'EXPORT INTO CSV 'nodelocal://0/%s/' FROM TABLE feature_flags': pq: parse "nodelocal://0/%s/": invalid URL escape "%s/"
    panic.go:522: -- test log scope end --
```

But note that in crdb if you try an export like this it fails with the same error anyways:
```
root@:26257/defaultdb> export into csv 'nodelocal://0/%s' from table foo;
ERROR: parse "nodelocal://0/%s": invalid URL escape "%s"
```

So i changed this `%s` to a normal path

alyshanjahani-crl commented:
bors r+

alyshanjahani-crl commented:
blathers backport 22.1 22.2
